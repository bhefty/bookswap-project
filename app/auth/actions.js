import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT,
  SET_SESSION
} from './constants'

export const loginRequest = () => ({
  type: LOGIN_REQUEST
})

export const loginSuccess = () => ({
  type: LOGIN_SUCCESS
})

export const loginFailure = (error) => ({
  type: LOGIN_FAILURE,
  error
})

export const logout = () => ({
  type: LOGOUT
})

export const setSession = (authResult) => ({
  type: SET_SESSION,
  authResult
})
