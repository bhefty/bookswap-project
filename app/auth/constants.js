export const LOGIN_REQUEST = 'bookswap-project/auth/LOGIN_REQUEST'
export const LOGIN_SUCCESS = 'bookswap-project/auth/LOGIN_SUCCESS'
export const LOGIN_FAILURE = 'bookswap-project/auth/LOGIN_FAILURE'
export const LOGOUT = 'bookswap-project/auth/LOGOUT'
export const SET_SESSION = 'bookswap-project/auth/SET_SESSION'
