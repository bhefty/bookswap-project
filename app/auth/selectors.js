import { createSelector } from 'reselect'

// If testing use the localStorage polyfill
let localStorage
/* istanbul ignore next */
if (global.process && process.env.NODE_ENV === 'test') {
  localStorage = require('localStorage')
} else {
  // If not, use the browser localStorage
  localStorage = global.window.localStorage
}

const selectAuth = state => state.get('auth')

const makeSelectAuthResult = () => createSelector(
  selectAuth,
  (authState) => authState.get('authResult')
)

const makeSelectIsAuthenticated = () => createSelector(
  selectAuth,
  (authState) => {
    const expiresAt = JSON.parse(localStorage.getItem('expires_at'))
    return (authState.get('isAuthenticated') && (new Date().getTime() < expiresAt))
  }
)

const makeSelectExpiryTime = () => {
  if (localStorage.getItem('expires_at')) {
    const expiresAt = JSON.parse(localStorage.getItem('expires_at'))
    return new Date().getTime() < expiresAt
  } else {
    return false
  }
}

export {
  selectAuth,
  makeSelectAuthResult,
  makeSelectIsAuthenticated,
  makeSelectExpiryTime
}
