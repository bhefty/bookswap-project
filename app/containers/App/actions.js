/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_TRANSLATION,
  LOAD_TRANSLATION_SUCCESS,
  LOAD_TRANSLATION_ERROR
} from './constants'

/**
 * Load the translation, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION
 */
export function loadTranslation () {
  return {
    type: LOAD_TRANSLATION
  }
}

/**
 * Dispatched when the translation is loaded by the request saga
 *
 * @param  {array} translation The translation data
 * @param  {string} text The current text
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION_SUCCESS passing the translation
 */
export function translationLoaded (translation, text) {
  return {
    type: LOAD_TRANSLATION_SUCCESS,
    translation,
    text
  }
}

/**
 * Dispatched when loading the translation fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_TRANSLATION_ERROR passing the error
 */
export function translationLoadingError (error) {
  return {
    type: LOAD_TRANSLATION_ERROR,
    error
  }
}
