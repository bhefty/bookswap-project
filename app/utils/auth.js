import auth0 from 'auth0-js'

export default class Auth {
  auth0 = new auth0.WebAuth({
    domain: 'bhefty.auth0.com',
    clientID: 'V6Suys2WmYNDVMjwXRDGIT1SPc0Rf48J',
    redirectUri: 'http://localhost:3000/callback',
    audience: 'https://bhefty.auth0.com/userinfo',
    responseType: 'token id_token',
    scope: 'openid profile'
  })

  constructor () {
    this.login = this.login.bind(this)
    this.handleAuthentication = this.handleAuthentication.bind(this)
  }

  login () {
    this.auth0.authorize()
  }

  handleAuthentication () {
    return new Promise((resolve, reject) => {
      this.auth0.parseHash((err, authResult) => {
        if (authResult && authResult.accessToken && authResult.idToken) {
          resolve(authResult)
        } else if (err) {
          reject(err)
        }
      })
    })
  }
}
