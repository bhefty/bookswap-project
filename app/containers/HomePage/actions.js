import * as types from './constants'

export const changeText = (text) => ({
  type: types.CHANGE_TEXT,
  text
})

export const increment = () => ({
  type: types.INCREMENT
})

export const decrement = () => ({
  type: types.DECREMENT
})
