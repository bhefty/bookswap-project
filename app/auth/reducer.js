import { fromJS } from 'immutable'

import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT,
  SET_SESSION
} from './constants'

import { makeSelectExpiryTime } from 'auth/selectors'

const isAuthenticated = makeSelectExpiryTime()

const initialState = fromJS({
  isLoggingIn: false,
  isAuthenticated,
  error: null,
  authResult: null
})

export default function authReducer (state = initialState, action) {
  switch (action.type) {
    case LOGIN_REQUEST:
      return state
        .set('isLoggingIn', true)
    case LOGIN_SUCCESS:
      return state
        .set('isLoggingIn', false)
        .set('isAuthenticated', true)
        .set('error', null)
    case LOGIN_FAILURE:
      return state
        .set('isLoggingIn', false)
        .set('error', action.error)
    case LOGOUT:
      return initialState
        .set('isAuthenticated', false)
    case SET_SESSION:
      return state
        .set('authResult', action.authResult)
    default:
      return state
  }
}
