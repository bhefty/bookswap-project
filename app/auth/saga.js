import { push } from 'react-router-redux'
import { take, call, put, all, fork, select } from 'redux-saga/effects'

import Auth from 'utils/auth'

import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT,
  SET_SESSION
} from './constants'

import { loginSuccess, loginFailure } from './actions'
import { makeSelectAuthResult } from './selectors'

const auth = new Auth()

let localStorage

// If testing use the localStorage polyfill
/* istanbul ignore next */
if (global.process && process.env.NODE_ENV === 'test') {
  localStorage = require('localStorage')
} else {
  // If not, use the browser localStorage
  localStorage = global.window.localStorage
}

// Login Request Handler
export function * loginRequestSaga () {
  try {
    yield call(auth.login)
  } catch (error) {
    yield put(loginFailure(error))
  }
}

// Set Session Handler
export function * setSessionSaga () {
  const authResult = yield select(makeSelectAuthResult())

  const expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime())
  localStorage.setItem('access_token', authResult.accessToken)
  localStorage.setItem('id_token', authResult.idToken)
  localStorage.setItem('expires_at', expiresAt)
  yield put(loginSuccess())
}

// Logout handler
export function * logoutSaga () {
  // Clear access token and ID token from local storage
  localStorage.removeItem('access_token')
  localStorage.removeItem('id_token')
  localStorage.removeItem('expires_at')
}

// Watcher functions
export function * watchLoginRequest () {
  yield take(LOGIN_REQUEST)
  yield call(loginRequestSaga)
}

export function * watchLoginSuccess () {
  yield take(LOGIN_SUCCESS)
  yield put(push('/dashboard'))
}

export function * watchLoginFailure () {
  yield take(LOGIN_FAILURE)
  yield put(push('/'))
}

export function * watchLogout () {
  yield take(LOGOUT)
  yield call(logoutSaga)
  yield put(push('/'))
}

export function * watchSetSession () {
  yield take(SET_SESSION)
  yield call(setSessionSaga)
}

// Root saga manages matcher lifecycle
export default function * authSagas () {
  yield all([
    fork(watchLoginRequest),
    fork(watchLoginSuccess),
    fork(watchLoginFailure),
    fork(watchLogout),
    fork(watchSetSession)
  ])
}
