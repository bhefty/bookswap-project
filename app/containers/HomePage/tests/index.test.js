import React from 'react'
import { shallow } from 'enzyme'

import { HomePage, mapDispatchToProps } from '../index'
import { loginRequest } from 'auth/actions'

describe('<HomePage />', () => {
  it('should render the page', () => {
    // const tree = renderer.create(
    //   <HomePage />
    // ).toJSON()
    // expect(tree).toMatchSnapshot()
    const renderedComponent = shallow(<HomePage isAuthenticated />)
    expect(renderedComponent.length).toEqual(1)
  })

  describe('_handleRoute', () => {
    it('should redirect to /dashboard if user is authenticated', () => {
      const mockPush = jest.fn()
      const renderedComponent = shallow(<HomePage isAuthenticated history={{push: mockPush}} />)
      renderedComponent.instance()._handleRoute()
      expect(mockPush).toHaveBeenCalled()
    })

    it('should dispatch loginRequest when user is not authenticated', () => {
      const mockLogin = jest.fn()
      const renderedComponent = shallow(<HomePage isAuthenticated={false} login={mockLogin} />)
      renderedComponent.instance()._handleRoute()
      expect(mockLogin).toHaveBeenCalled()
    })
  })

  describe('mapDispatchToProps', () => {
    describe('login', () => {
      it('should be injected', () => {
        const dispatch = jest.fn()
        const result = mapDispatchToProps(dispatch)
        expect(result.login).toBeDefined()
      })

      it('should dispatch loginRequest when called', () => {
        const dispatch = jest.fn()
        const result = mapDispatchToProps(dispatch)
        result.login()
        expect(dispatch).toHaveBeenCalledWith(loginRequest())
      })
    })
  })
})
