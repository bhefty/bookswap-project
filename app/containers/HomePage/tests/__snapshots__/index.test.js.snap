// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render the page 1`] = `
<article>
  <div
    className="HomePage__Splash-eQGvGD icCRCs"
  >
    <main>
      <img
        alt="Bookswap Logo"
        src="IMAGE_MOCK"
      />
      <h1>
        Bookswap
      </h1>
      <h3>
        Quickly and easily trade books with like-minded friends.
      </h3>
      <button
        onClick={[Function]}
      >
        Get Started
        <svg
          fill="#FFFFFF"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
          />
          <path
            d="M0-.25h24v24H0z"
            fill="none"
          />
        </svg>
      </button>
    </main>
  </div>
</article>
`;
